pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -3411,15 +3412,18 @@ private long readLongValue(byte[] bytes, int oid, long minVal, long maxVal, Stri
         val = ByteConverter.int8(bytes, 0);
         break;
       case Oid.FLOAT4:
-        val = (long) ByteConverter.float4(bytes, 0);
+        bd = BigDecimal.valueOf(ByteConverter.float4(bytes, 0));
```

Comment:
Do we really need to go with `BigDecimal` for `FLOAT4`, `FLOAT8`?

Reply:
I imagine it simplifies the logic

Reply:
It copies how the string parsing logic works.
Nothing jumped out at me as obvious way to check for overflow with primitive types

Reply:
Should probably use constructor rather than valueOf though. Exact binary value is appropriate.

Reply:
Can you elaborate which overflow do you mean?
Is there anything wrong with comparing `float` with `Long.MAX_VALUE`?

Reply:
I wonder if this change has performance implications

Reply:
comparing a float (or double) with a long results in the long being converted to a float (or double), and I do not know what that value will be, as Long.MAX_VALUE (and MIN_VALUE) are outside range either float or double can maintain perfect integer precision. So some rounding will occur.
I did just change to directly cast to long if the value is within the range of an int.

Reply:
Well, there's `StrictMath.nextDown` and `StrictMath.nextUp`, so we can have exact floating constants representing max value and min value.

For instance, `(float) 9223371487098961920L` is the exact largest `float` that still fits within `Long.MAX_VALUE`. Th next float is `9223372036854775808` which exceeds `Long.MAX_VALUE`.

Reply:
In this latest commit, I created constants for min and max long in range of floats. For values in that range, we can just cast to long. Outside that range, we use BigDecimal.

Reply:
@vlsi thoughts on current impl?